opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/LuceneQuery.java

```diff
@@ -102,7 +106,7 @@ private boolean literalExpressionWrappedByCast(FunctionExpression func) {
    * @return query
    */
   public QueryBuilder build(FunctionExpression func) {
-    ReferenceExpression ref = (ReferenceExpression) func.getArguments().get(0);
+    this.ref = (ReferenceExpression) func.getArguments().get(0);
```

Comment:
Will this class be used in multi-threads? `ref` in `LuceneQuery` is shared member in multi-threads.
I think you can change to:
1. Add a parameter `ReferenceExpression ref` in `getParsedDateTime()`
```java
private ZonedDateTime getParsedDateTime(LiteralExpression expr, ReferenceExpression ref)
```
2. Change `cast()` method to
```java
  private ExprValue cast(FunctionExpression castFunction, ReferenceExpression ref) {
    return castMap
        .get(castFunction.getFunctionName())
        .apply((LiteralExpression) castFunction.getArguments().get(0), ref);
  }
```
3. Change the `castMap` with BiFunction
```java
private final Map<FunctionName, BiFunction<LiteralExpression, ReferenceExpression, ExprValue>> castMap =
         ...
         .put(
              BuiltinFunctionName.CAST_TO_DATE.getName(),
              (expr, ref) -> {
                if (expr.type().equals(ExprCoreType.STRING)) {
                  ZonedDateTime zonedDateTime = getParsedDateTime(expr, ref);
                  if (zonedDateTime != null) {
                    return new ExprDateValue(zonedDateTime.toLocalDate());
                  }
                  return new ExprDateValue(expr.valueOf().stringValue());
                } else {
                  return new ExprDateValue(expr.valueOf().dateValue());
                }
              })
```

Reply:
Since only `ref.type()` used in `getParsedDateTime`,
Pass a `ExprType` parameter instead of a `ReferenceExpression` in above solution would be better in case serialize/deserialize problem in lambda function.

Reply:
Good call! Updated